pipeline_dp/pipeline_operations.py

```diff
@@ -323,3 +329,172 @@ def count_per_element(self, col, stage_name: typing.Optional[str] = None):
 
     def reduce_accumulators_per_key(self, col, stage_name: str = None):
         raise NotImplementedError()
+
+# workaround for passing lambda functions to multiprocessing
+# according to https://medium.com/@yasufumy/python-multiprocessing-c6d54107dd55
+_pool_current_func = None
+def _pool_worker_init(func):
+    global _pool_current_func
+    _pool_current_func = func
+
+def _pool_worker(row):
+    return _pool_current_func(row)
+
+class MultiProcLocalPipelineOperations(PipelineOperations):
+    def __init__(self, n_jobs: typing.Optional[int]=None,
+                chunksize: int=1,
+                ordered=True, 
+                **pool_kwargs):
+        if n_jobs is None:
+            n_jobs = os.cpu_count()
+        self.n_jobs = n_jobs
+        self.chunksize = chunksize
+        self.ordered = ordered
+        self.pool_kwargs = pool_kwargs
+        
+        if ordered:
+            self._pool_map_fn = mp.pool.Pool.map
+        else:
+            self._pool_map_fn = mp.pool.Pool.imap_unordered
+
+    def map(self, col, fn, stage_name: typing.Optional[str]=None):
+        with mp.Pool(self.n_jobs, initializer=_pool_worker_init, initargs=(fn,)) as pool:
+            return self._pool_map_fn(pool, _pool_worker, col, chunksize=self.chunksize)
+
+    def flat_map(self, col, fn, stage_name: typing.Optional[str]=None):
+        return (e for x in self.map(col, fn, stage_name) for e in x)
+
+    def map_tuple(self, col, fn, stage_name: typing.Optional[str]=None):
+        # def mapped_fn(captures: typing.Tuple[typing.Callable], row):
+        #     func, = captures
+        #     return func(*row)
+        # mapped_fn = partial(mapped_fn, (fn, ))
+        
+        return self.map(col, lambda row: fn(*row), stage_name)
+
+    def map_values(self, col, fn, stage_name: typing.Optional[str]=None):
+        # def mapped_fn(captures: typing.Tuple[typing.Callable], row):
+        #     func, = captures
+        #     return row[0], func(row[1])
+        # mapped_fn = partial(mapped_fn, (fn, ))
+        return self.map(col, lambda x: (x[0], fn(x[1])), stage_name)
+
+    def group_by_key(self, col, stage_name: typing.Optional[str]=None):
+        # NOTE - this cannot be implemented in an ordered manner without (almost) serial execution!
+        #   both keys and groups will be out of order
+        keys = set(self.keys(col))
+        with mp.Manager() as manager:
+            results_dict = manager.dict({
```

Comment:
Is manager.dict protected somehow against race conditions?

Reply:
[Yes](https://docs.python.org/3/library/multiprocessing.html#sharing-state-between-processes), it's a proxy for the `dict` object. Internally it's a `mp.Queue` that feeds requests for `dict.__setitem__` to a central server process holding the actual `dict` object.